client/gutenboarding/store/resolvers.ts

```diff
@@ -0,0 +1,31 @@
+/**
+ * External dependencies
+ */
+import { addQueryArgs } from '@wordpress/url';
+
+/**
+ * Internal dependencies
+ */
+import * as selectors from './selectors';
+import { apiFetch, dispatch } from '@wordpress/data-controls';
+import { STORE_KEY } from './constants';
+
+type ResolverParameters< S extends ( state: any, ...args: any ) => any > = S extends (
+	state: any,
+	...args: infer P
+) => any
+	? P
+	: never;
```

Comment:
This seems super-abstract, and generic (and I'm having a hard time grokking it :sweat_smile: )

Is this not available from the DefinitelyTyped lib? Would it make sense to add it there?

Reply:
I took a look at [`utility-types`](https://github.com/piotrwitek/utility-types) and didn't find anything to do this. A documentation comment should be added 👍 

In https://github.com/Automattic/wp-calypso/pull/37165 I reused this and used less specific names, I think we should go with `TailParameters`, matching common `head`/`tail` functions to access `[ head, ...tail ]`. I'll pull those changes here.

What this does is infer the type of the parameters for a function skipping the first. This is handy because it allows us to type some things related to selectors that would otherwise be difficult in `@wordpress/data`. Selector functions accept the store `state` as a first parameter, but are never passed the store `state` when they're invoked. Resolvers also follow this pattern, their parameters correspond to the selector of the same name, with the initial `state` argument dropped.

It's a bit like pattern matching, here's a way to think about it:

- We provide `TailParameters<S>`
- Does `S` have a shape like `( state: any, ...args: P ) => any`? (we have a conditional type here, so it branches)
- In the success branch, `infer` in `( state: any, ...args: infer P )` says "TypeScript, you know the shape so figure out that piece." This is how we extract a piece of the shape of that type, like in pattern matching.
- The fail branch is [`never`](https://www.typescriptlang.org/docs/handbook/basic-types.html#never), a type we should never arrive at.

The way `infer` works requires a conditional type which are the things that look like a ternary. Because this is the way to "pattern-match" and extract pieces of more complex types, you'll see this pattern over and over.

Reply:
I agree it looks difficult to understand when just reading code. I like the idea of renaming to `TailParameters`.

Would it help to expand it to more lines of code to increase readability? Sometimes code just suffers "too smartly written" symptom (a good kind of problem to have!) and expanding compact code often helps. I don't know if it's easy to do here tho.

Reply:
- This is a complex type, but I believe its complexity is essential.
- I tried a strategy to break it down, but it is difficult or impossible to get results that are as good as this.
- It's using several of the more advanced features of TypeScript, so it's unsurprising that it's tricky to understand.
- It's not a type that's intended to be heavily used. It does provide type-level functionality so that we have access to very informative types.

This is a sort of "behind the scenes" piece of functionality I don't think it's worth spending a lot of time on. We can add a docstring to clarify what it does, stick it in some application-level type helpers if we're interested, but mostly forget it exists.

Reply:
> The way `infer` works requires a conditional type which are the things that look like a ternary. Because this is the way to "pattern-match" and extract pieces of more complex types, you'll see this pattern over and over.

:thinking: And there's no shorthand for that, if it's recurring? Or a different way to pattern-match when... we can assume that `infer` won't fail? (A bit hand-wavy here, sorry, definitely not in my element here :sweat_smile: -- just trying to, err, infer things at a somewhat meta level.)

Reply:
Operating on types we're already at a meta level 😁 

TypeScript calls these sort of "type functions" [_mapped types_](https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types):

> TypeScript provides a way to create new types based on old types — mapped types.

This combines with conditional types and [`infer` declarations](https://www.typescriptlang.org/docs/handbook/advanced-types.html#type-inference-in-conditional-types) to allow us to do this sort of "type pattern matching." Given a type with a shape, we can extract a piece of it.

I don't think there's any shortcuts we can make here, operating at the type level is a bit different, remember we don't have values and can't evaluate things in the way we're accustomed to with runtime JavaScript.

It may seem like something we could trim down, but I just don't think it's possible within the constraints of this type system. And yes, this exact pattern appears [again](https://github.com/microsoft/TypeScript/blob/87cc8c4af8fb2adcde0dfe92cb6f4703236b3291/lib/lib.es5.d.ts#L1506) and [again](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5c197f5c3974464c81bdd99a453c8a7cabe9a96a/types/react/index.d.ts#L828) and [again](https://github.com/reduxjs/redux/blob/9d3273846aa8906d38890c410b62fb09a4992018/src/types/reducers.ts#L48-L53). If there were a better solution at this time, I'm sure core or a utility library would already be providing it 🙂